src/zenml/zen_server/rbac/utils.py

```diff
@@ -0,0 +1,569 @@
+            model=value, action=Action.READ
+        ):
+            return dehydrate_response_model(value, permissions=permissions)
+        else:
+            return get_permission_denied_model(value)
+    elif isinstance(value, BaseModel):
+        return dehydrate_response_model(value, permissions=permissions)
+    elif isinstance(value, Dict):
+        return {
+            k: _dehydrate_value(v, permissions=permissions)
+            for k, v in value.items()
+        }
+    elif isinstance(value, (List, Set, tuple)):
+        type_ = type(value)
+        return type_(
+            _dehydrate_value(v, permissions=permissions) for v in value
+        )
+    else:
+        return value
+
+
+def has_permissions_for_model(model: AnyResponseModel, action: str) -> bool:
+    """If the active user has permissions to perform the action on the model.
+
+    Args:
+        model: The model the user wants to perform the action on.
+        action: The action the user wants to perform.
+
+    Returns:
+        If the active user has permissions to perform the action on the model.
+    """
+    if is_owned_by_authenticated_user(model):
+        return True
+
+    try:
+        verify_permission_for_model(model=model, action=action)
+        return True
+    except HTTPException:
+        return False
+
+
+def get_permission_denied_model(model: AnyResponseModel) -> AnyResponseModel:
+    """Get a model to return in case of missing read permissions.
+
+    Args:
+        model: The original model.
+
+    Returns:
+        The permission denied model.
+    """
+    if isinstance(model, BaseResponse):
+        return cast(AnyResponseModel, get_permission_denied_model_v2(model))
+    else:
+        return cast(
+            AnyResponseModel,
+            get_permission_denied_model_v1(cast(BaseResponseModel, model)),
+        )
+
+
+def get_permission_denied_model_v2(
+    model: AnyNewResponseModel,
+) -> AnyNewResponseModel:
+    """Get a V2 model to return in case of missing read permissions.
+
+    This function removes the body and metadata of the model.
+
+    Args:
+        model: The original model.
+
+    Returns:
+        The model with body and metadata removed.
+    """
+    return model.copy(
+        exclude={"body", "metadata"}, update={"permission_denied": True}
+    )
+
+
+def get_permission_denied_model_v1(
+    model: AnyOldResponseModel, keep_id: bool = True, keep_name: bool = True
+) -> AnyOldResponseModel:
+    """Get a V1 model to return in case of missing read permissions.
+
+    This function replaces all attributes except name and ID in the given model.
+
+    Args:
+        model: The original model.
+        keep_id: If `True`, the model ID will not be replaced.
+        keep_name: If `True`, the model name will not be replaced.
+
+    Returns:
+        The model with attribute values replaced by default values.
+    """
+    values = {}
+
+    for field_name, field in model.__fields__.items():
```

Comment:
Same here: should you exclude private fields, just in case a model already uses or decides to use that ?

Reply:
See above, not included